### PR TITLE
policygen: support reading state files from gcs

### LINKED ISSUE #206: Support reading state files from a GCS bucket.

### PR DESCRIPTION
**Pull request by xingao267**
Fixes #206 

TESTED=using our testing devops project

### CODE REVIEW

File: internal/policygen/load.go [hunk 1]
```diff
@@ -94,33 +100,35 @@ func loadConfig(path string) (*config, error) {
...
 }
 
-// loadResources loads Terraform state resources from the given path. If the path is a single
-// file, it loads resouces from it. If the path is a directory, it walks the directory
-// recursively and loads resources from each .tfstate file.
+// loadResources loads Terraform state resources from the given path.
+// - If the path is a single local file, it loads resouces from it.
+// - If the path is a local directory, it walks the directory recursively and loads resources from each .tfstate file.
+// - If the path is a GCS bucket (indicated by 'gs://' prefix), it walks the bucket recursively and loads resources from each .tfstate file.
+//   It currently ignores the file/dir path inside the bucket and considers all .tfstate file in the bucket.
```
Comment (MartinPetkov) [inline internal/policygen/load.go:107]: Is it difficult to consider the file or dir in the bucket?

Comment (xingao267) [reply internal/policygen/load.go]: It's not difficult but just will make the code a bit messy and we currently don't need that. We can add that when there is a FR for that.

File: internal/policygen/load.go [hunk 2]
```diff
@@ -91,36 +97,39 @@ func loadConfig(path string) (*config, error) {
...
 	if !fi.IsDir() {
-		resources, err := terraform.ResourcesFromState(path)
-		if err != nil {
-			return nil, fmt.Errorf("read resources from Terraform state file %q: %v", path, err)
-		}
-		return resources, nil
+		return loadResourcesFromSingleLocalFile(path)
 	}
 
 	var allResources []*states.Resource
```
Comment (umairidris) [inline internal/policygen/load.go:124]: Consider moving this to a function e.g. loadResourcesFromDir.

File: internal/policygen/load.go (hunk 1 above)
Comment (MartinPetkov) [reply internal/policygen/load.go]: sg

File: internal/policygen/load.go [hunk 3]
```diff
@@ -91,36 +97,39 @@ func loadConfig(path string) (*config, error) {
...
 	fn := func(path string, _ os.FileInfo, _ error) error {
```
Comment (umairidris) [inline internal/policygen/load.go:125]: Check error?

File: internal/policygen/load.go [hunk 4]
```diff
@@ -130,5 +139,66 @@ func loadResources(path string) ([]*states.Resource, error) {
...
+	}
+	return resources, nil
+}
+
+func loadResourcesFromCloudStorageBucket(path string) ([]*states.Resource, error) {
+	// Trim the 'gs://' prefix and split the path into the bucket name and cloud storage file path.
+	bucketName := strings.SplitN(strings.TrimPrefix(path, "gs://"), "/", 2)[0]
+	log.Printf("Reading state files from Cloud Storage bucket %q", bucketName)
+
+	ctx := context.Background()
```
Comment (umairidris) [inline internal/policygen/load.go:159]: Pass context from top level caller

File: internal/policygen/load.go (hunk 3 above)
Comment (xingao267) [reply internal/policygen/load.go:125]: Done

**Pull request merged**